cdap-provisioner-spi/src/main/java/co/cask/cdap/provisioner/spi/ProvisionerProperty.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.provisioner.spi;
+
+import java.util.Objects;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Property of a Provisioner. Includes optional restrictions on the property value and range.
+ */
+public class ProvisionerProperty {
+  private final String name;
+  private final String label;
+  private final String description;
+  private final String type;
+  private final Set<String> values;
+  private final Range range;
+
+  public ProvisionerProperty(String name, String label, String description, String type,
+                             @Nullable Set<String> values, @Nullable Range range) {
```

Comment:
I think instead of having optional `values` and `range` (which they are mutually exclusive, I believe), it is better to have sub-classes to impose restrictions on the property value, and have methods exposed for the validation.